tests/src/core/testqgsjsonutils.cpp

```diff
@@ -91,19 +91,17 @@ void TestQgsJsonUtils::testJsonArray()
   QCOMPARE( QgsJsonUtils::parseArray( R"([])", QVariant::Int ), QVariantList() );
   QCOMPARE( QgsJsonUtils::parseArray( "", QVariant::Int ), QVariantList() );
   // Nulls
-  QCOMPARE( QgsJsonUtils::parseArray( R"([null, null])" ), QVariantList() << QVariant() << QVariant() );
-  QVERIFY( QVariant().isValid() );
   for ( const QVariant &value : QgsJsonUtils::parseArray( R"([null, null])" ) )
   {
     QVERIFY( value.isNull() );
-    // This would fail because the expected type is not set and an invalid QVariant is created:
-    // QVERIFY( value.isValid() );
+    QVERIFY( value.isValid() );
+    QCOMPARE( value, QVariant( QVariant::Type::Int ) );
```

Comment:
Nice!

Now really out of curiosity:
I normally don't test for API behavior which is more a "side-effect" that could potentially change any time (let's assume it's replaced by `QVariant::Type::Dynamic` in the future). How do you feel about this kind of tests?

Reply:
Sorry, I'm not sure I understand :) 
Do you mean I should not test this case or that there is a better (more resilient) way to test it?

Reply:
The first one. But I didn't mean "you shouldn't" and was more out for an answer to the question "what advantage do we expect from this test".

Reply:
ok, maybe an obvious answer but the test is IMO important to test that the method behaves as expected. If I got your question right, we might better define (in `qgis.h` maybe) a `QVariant::Type::Something`  that we use as a fallback in case we want something to represent a  `NULL` but not invalid (i.e. undefined), and the use that alias in the tests (and in the code of course). 
Or perhaps go the `QgsVariant` way to handle `NULL` explicitely.

Reply:
Then the question is, is returning an int "expected" or just "a side-effect". I think that no code consuming this API should rely on this fact (it should just test for `isValid()`, `isNull()` or rely on other sources of type information but not the NULL value itself. Hence the question: what's the advantage of testing that workarounds are kept the way they are. I hope I could explain a bit better :)

In short: Implementation-wise I would not change anything (I don't see a better approach), but I wonder if we should add tests to ensure leftovers of workarounds are "stable".

Reply:
Ok, got it finally  :)
I agree that it should probably not be tested, isNull and isValid should be sufficient, I'll try to keep that in mind next time.
I find myself writing tests with two different goals, one is to check that the method works as expected (the Int here) and the other one is more API oriented and it is that it won't break in the future (kind of policy/insurance), I think that I should remove or comment the tests o in the first category, but it's not always easy to tell them apart.

Reply:
Sounds good. I was just thinking loud here.